From dc3f9fee810bb06d9351d784bf0fbddd8e5a0a55 Mon Sep 17 00:00:00 2001
From: John Crispin <john@phrozen.org>
Date: Tue, 5 Oct 2021 21:09:37 -0700
Subject: [PATCH 01/14] mac80211: MBSSID and EMA beacon handling in AP mode

Add new fields in struct beacon_data to store all MBSSID elements.

For a non-EMA AP, generate a single beacon template which includes all
MBSSID elements.
For an EMA AP, generate multiple beacon templates, each including
a single MBSSID element. EMA profile periodicity equals the count of
elements.

Move CSA offset to reflect the MBSSID element length.

Add new APIs for the drivers to retrieve MBSSID and EMA beacons:
- ieee80211_beacon_get_template_ema_index() - Generate the EMA beacon
  which includes the multiple BSSID element at the given index.
  Drivers can use this function in a loop as well NULL is returned.
- ieee80211_beacon_get_template_ema_list() - Generate all EMA templates.
  Drivers must call ieee80211_beacon_free_ema_list() to free the memory.
  No change in the prototype for the existing API
  ieee80211_beacon_get_template() which should be used for non-EMA cases,
  both with and without multiple BSSID enabled.

Signed-off-by: John Crispin <john@phrozen.org>
Co-developed-by: Aloka Dixit <alokad@codeaurora.org>
Signed-off-by: Aloka Dixit <alokad@codeaurora.org>
---
 include/net/mac80211.h     |  92 ++++++++++++++++++++
 net/mac80211/cfg.c         | 172 ++++++++++++++++++++++++++++++-------
 net/mac80211/ieee80211_i.h |   1 +
 net/mac80211/tx.c          | 149 ++++++++++++++++++++++++++++++--
 4 files changed, 377 insertions(+), 37 deletions(-)

--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -5408,25 +5408,48 @@ ieee80211_beacon_get_template(struct iee
 			      unsigned int link_id);
 
 /**
+ * enum ieee80211_bcn_tmpl_ema - EMA beacon generation type
+ *
+ * When enhanced multi-BSSID advertisements (EMA) mode is enabled, the
+ * non-transmitting profiles from the multiple BSSID set are split into more
+ * than one multiple BSSID elements if required. Each EMA beacon includes only
+ * one element to reduce the total size. The number of beacons required to
+ * cover all profiles is called as the profile periodicity of the set.
+ *
+ * @IEEE80211_BCN_EMA_DISABLED: Used when EMA is disabled. Only one beacon
+ *	template will be generated. This beacon will also includes all
+ *	multiple BSSID elements if present.
+ * @IEEE80211_BCN_EMA_BASE: Used when EMA is enabled. Beacon template includes
+ *	the multiple MBSSID element at a specified index which should be set
+ *	to a value more than or equal to IEEE80211_BCN_EMA_BASE.
+ */
+enum ieee80211_bcn_tmpl_ema {
+	IEEE80211_BCN_EMA_DISABLED = -1,
+	IEEE80211_BCN_EMA_BASE = 0,
+};
+
+/**
  * ieee80211_beacon_get_template_ema_index - EMA beacon template generation
+ *	function for drivers using the sw offload path.
  * @hw: pointer obtained from ieee80211_alloc_hw().
  * @vif: &struct ieee80211_vif pointer from the add_interface callback.
  * @offs: &struct ieee80211_mutable_offsets pointer to struct that will
  *	receive the offsets that may be updated by the driver.
- * @link_id: the link id to which the beacon belongs (or 0 for a non-MLD AP).
- * @ema_index: index of the beacon in the EMA set.
+ * @ema_index: index of the beacon in the EMA set, must be more than or equal
+ *	to 0.
  *
  * This function follows the same rules as ieee80211_beacon_get_template()
  * but returns a beacon template which includes multiple BSSID element at the
  * requested index.
  *
- * Return: The beacon template. %NULL indicates the end of EMA templates.
+ * Return: The beacon template. %NULL indicates the end of EMA beacon templates.
  */
 struct sk_buff *
 ieee80211_beacon_get_template_ema_index(struct ieee80211_hw *hw,
 					struct ieee80211_vif *vif,
+		       		unsigned int link_id,
 					struct ieee80211_mutable_offsets *offs,
-					unsigned int link_id, u8 ema_index);
+					u8 ema_index);
 
 /**
  * struct ieee80211_ema_beacons - List of EMA beacons
@@ -5447,9 +5470,10 @@ struct ieee80211_ema_beacons {
 
 /**
  * ieee80211_beacon_get_template_ema_list - EMA beacon template generation
+ *	function for drivers using the hw offload.
  * @hw: pointer obtained from ieee80211_alloc_hw().
  * @vif: &struct ieee80211_vif pointer from the add_interface callback.
- * @link_id: the link id to which the beacon belongs (or 0 for a non-MLD AP)
+ * @ema_beacons: list of EMA beacons of type &struct ieee80211_ema_beacons.
  *
  * This function follows the same rules as ieee80211_beacon_get_template()
  * but allocates and returns a pointer to list of all beacon templates required
@@ -5464,7 +5488,7 @@ struct ieee80211_ema_beacons {
 struct ieee80211_ema_beacons *
 ieee80211_beacon_get_template_ema_list(struct ieee80211_hw *hw,
 				       struct ieee80211_vif *vif,
-				       unsigned int link_id);
+		       		   unsigned int link_id);
 
 /**
  * ieee80211_beacon_free_ema_list - free an EMA beacon template list
@@ -5476,6 +5500,7 @@ ieee80211_beacon_get_template_ema_list(s
 void ieee80211_beacon_free_ema_list(struct ieee80211_ema_beacons *ema_beacons);
 
 /**
+
  * ieee80211_beacon_get_tim - beacon generation function
  * @hw: pointer obtained from ieee80211_alloc_hw().
  * @vif: &struct ieee80211_vif pointer from the add_interface callback.
--- a/net/mac80211/ieee80211_i.h
+++ b/net/mac80211/ieee80211_i.h
@@ -1210,9 +1210,12 @@ ieee80211_get_mbssid_beacon_len(struct c
 {
 	int len = 0;
 
-	if (!elems || !elems->cnt || i > elems->cnt)
+	if (!elems || !elems->cnt || i >= elems->cnt)
 		return 0;
 
+	if (i >= IEEE80211_BCN_EMA_BASE)
+		return elems->elem[i].len;
+
 	if (i < elems->cnt) {
 		len = elems->elem[i].len;
 		if (rnr_elems) {
--- a/net/mac80211/tx.c
+++ b/net/mac80211/tx.c
@@ -5533,9 +5533,15 @@ ieee80211_beacon_add_mbssid(struct sk_bu
 			    u8 i)
 {
 	if (!beacon->mbssid_ies || !beacon->mbssid_ies->cnt ||
-	    i > beacon->mbssid_ies->cnt)
+	    i >= beacon->mbssid_ies->cnt)
 		return;
 
+	if (i >= IEEE80211_BCN_EMA_BASE) {
+		skb_put_data(skb, beacon->mbssid_ies->elem[i].data,
+			     beacon->mbssid_ies->elem[i].len);
+		return;
+	}
+
 	if (i < beacon->mbssid_ies->cnt) {
 		skb_put_data(skb, beacon->mbssid_ies->elem[i].data,
 			     beacon->mbssid_ies->elem[i].len);
@@ -5815,8 +5821,9 @@ EXPORT_SYMBOL(ieee80211_beacon_get_templ
 struct sk_buff *
 ieee80211_beacon_get_template_ema_index(struct ieee80211_hw *hw,
 					struct ieee80211_vif *vif,
+				unsigned int link_id,
 					struct ieee80211_mutable_offsets *offs,
-					unsigned int link_id, u8 ema_index)
+					u8 ema_index)
 {
 	return __ieee80211_beacon_get(hw, vif, offs, true, link_id, ema_index,
 				      NULL);
@@ -5843,9 +5850,8 @@ ieee80211_beacon_get_template_ema_list(s
 				       unsigned int link_id)
 {
 	struct ieee80211_ema_beacons *ema_beacons = NULL;
-
-	WARN_ON(__ieee80211_beacon_get(hw, vif, NULL, true, link_id, 0,
-				       &ema_beacons));
+	(void)__ieee80211_beacon_get(hw, vif, NULL, true, link_id,
+			       IEEE80211_BCN_EMA_BASE, &ema_beacons);
 
 	return ema_beacons;
 }
@@ -5859,7 +5865,7 @@ struct sk_buff *ieee80211_beacon_get_tim
 	struct ieee80211_mutable_offsets offs = {};
 	struct sk_buff *bcn = __ieee80211_beacon_get(hw, vif, &offs, false,
 						     link_id,
-						     IEEE80211_INCLUDE_ALL_MBSSID_ELEMS,
+						     IEEE80211_BCN_EMA_DISABLED,
 						     NULL);
 	struct sk_buff *copy;
 	int shift;
